### ISSUE TITLE
Feature request: Console next to the editor in the demo website (for education)

### ISSUE THREAD

**Issue by MayeulC** [brython-dev/brython#1131]
## Context
One of my acquaintances is a high-school maths teacher, and he wants a no-install way for students to try their python code, while not requiring hosting a solution.
The interface has to be very simple, and the feature set basic as to not confuse students and focus on the lesson that has quite a small number of hours (iodide allowing to mix languages is already a barrier to entry, for instance).
Another requirement eliminating many possible solution is that the python version has to be quite recent (fstrings).

## Feedback

I suggested Brython, and he was pretty happy with it. However, he let me know that he would have preferred a spyder-style interface with the editor code being executed in an interactive console, to let the user interactively explore the variables after the program has ran.

Would it be possible to implement this in the [editor](https://brython.info/tests/editor.html) page? I had a look at it myself, but could not find a way to do so easily.
Running the script in the editor should probably import the console environment, and the console should give an interactive access to that environment after the script has ran.

## Further thoughts

Still in the same context, I think it would be valuable to teachers if there was an example html page (self-contained, or referencing external scripts) that could be downloaded, and easily modified to pre-fill the editor with some example code. It could provide classrooms with a one-html-page-per-exercise approach.

Likewise, I tried to re-implement a basic editor myself, but couldn't figure which javascript files were necessary to allow loading various modules.

Even better if the students could save their work back (just ctrl+s the web page would be best, but looks like browser support might differ quite a bit, so maybe a download/upload script button).

**Comment by PierreQuentel** [comment]
Bonjour Mayeul,

This is a very interesting idea and potentially a major development path for Brython. I have been playing with this kind of project for a while but I much prefer working on a real use case.

I have prototyped something that is hopefully a step in the direction you and the maths teacher require. I have uploaded it on the Brython site : [http://brython.info/gallery/console](http://brython.info/gallery/console). It includes: 

- a Python code editor with syntax highlighting, with the latest version of the Ace editor. The Python code can be loaded from a file in the user's file system, and saved in the browser's download directory
- a console used as the standard output and error streams when running Python scripts. When the code has finished running, the console opens an interactive session, which runs in the namespace defined by the script

The interface is in English by default, but if the browser has language set to French, the interface is translated.

I have not tested it completely but I thought I would show the result at this early stage to get a first feedback.

Merci !
Pierre

**Comment by MayeulC** [comment]
Bonjour Pierre :wink: 

Wow, thank you very much, this is quite interesting, and I will share it right away with the interested party.

I spent a few minutes playing with it, and it looks fairly functional from my side. I hope to gather some more constructive feedback on this, as it could indeed be an interesting development path.

Many mercis to you as well,

Mayeul

**Comment by PierreQuentel** [comment]
I have uploaded a new version. New features:

- if loaded with a query string `?file=script.py`, the script (in the same directory as the html page) is loaded in the editor. Hopefully this is a solution for your request to prefill the editor with an example, without having to save an HTML page for each example
- the program can manage several scripts at a time. A panel on the left shows the files currently managed; the currently edited file is highlighted. The panel is updated when a new file is opened. Clicking on a file name loads its content in the editor. The cursor position when it was last edited is saved
- instead of two buttons, "Save" and "Save as...", there is a single button "Save". It is intended to be clicked with the _right_ click; the right click menu has a line "Enregistrer la cible du lien sous..." or equivalent. This allows the user to select the directory where the file will be saved ; in the previous version, it was saved in the browser's download directory.

I didn't mention it before, but the console at the right can be used without running a script loaded in the editor: hitting the Enter key shows the interactive prompt (`>>>`)

**Comment by MayeulC** [comment]
Hello, thank you for your continued work on this! The teacher in question seemed pretty happy with this, though he noted that he wouldn't be able to try it with students for a while because of the summer vacations.

Here is some quick feedback on your work, which looks quite nice already (I went a bit beyond, don't feel like you have to implement everything I am asking for the way I ask it :slightly_smiling_face: ).


> I have uploaded a new version. New features:
> 
>   * if loaded with a query string `?file=script.py`, the script (in the same directory as the html page) is loaded in the editor. Hopefully this is a solution for your request to prefill the editor with an example, without having to save an HTML page for each example

Just to be clear: this request was my own, as I can see it being useful if distributed on a usb stick or similar. It looks quite good indeed, and makes me want to try it within iframes :)
Would it be simple to add a few scripts as parameters to preload them in the script list?

>   * the program can manage several scripts at a time. A panel on the left shows the files currently managed; the currently edited file is highlighted. The panel is updated when a new file is opened. Clicking on a file name loads its content in the editor. The cursor position when it was last edited is saved

Well, I think the UI is a bit confusing, as it appears after loading a file. Maybe add a `new script` button in that list?

>   * instead of two buttons, "Save" and "Save as...", there is a single button "Save". It is intended to be clicked with the _right_ click; the right click menu has a line "Enregistrer la cible du lien sous..." or equivalent. This allows the user to select the directory where the file will be saved ; in the previous version, it was saved in the browser's download directory.

Hmm, it doesn't look functional on my system (Firefox/Linux): by doing this, I obtained a web page, without the script, and that was unable to run python scripts.
To be fair, I think it is a bit confusing, as it is the first time I see that kind of pattern being used. I personally set my browsers to ask for a location before saving, but I get that not everyone does it.

I'm not sure what the correct solution is there. Maybe have some kind of old-school "statusbar" giving tips about what is under the mouse and the possible actions? ("tip: right click -> save as to select a location).

> I didn't mention it before, but the console at the right can be used without running a script loaded in the editor: hitting the Enter key shows the interactive prompt (`>>>`)

Thanks for mentioning this :) Of course, a placeholder string would be much welcome for UX :)

-----
Workflow ideas:

In the future, I think local storage could be used for some kind of... well, local storage (giving access to the interpreter even offline, as well as the previous scripts), and saving everything would give a zip file with the python scripts.

I would add a list of scripts, with:
* A ":heavy_plus_sign:" icon to add a new script (maybe with "create new script" besides it) 
* A way to rename a script
* When the list is empty (first time loading), give the current script a name: "new script.py"
* For each script, 
   * A ":x:" delete button (with confirmation...)
   * A " :arrow_down:" button to download
   * A ":ballot_box_with_check:" checkbox to perform the two actions above on multiple scripts

**Comment by PierreQuentel** [comment]
I don't forget this issue, but put it at the bottom of the todo list until the end of the holidays ;-)

Since this is not directly part of Brython I have opened a new project called [levraoueg](https://github.com/brython-dev/levraoueg) (library, in Breton - the name is likely to change) where I have included some of the suggestions above. One of the features is that the code is saved in an indexedDB database instead of the file system, or of local storage, which is limited in size.

For the discussion about this tool the best is to use its own issue tracker.
